pkg/utils/finalizer.go

```diff
@@ -25,25 +25,37 @@ type Finalizer struct {
 	pending []func() error
 }
 
-func (f *Finalizer) With(fi func() error) {
+func (f *Finalizer) Lock(locker sync.Locker) *Finalizer {
+	locker.Lock()
+	return f.WithVoid(locker.Unlock)
+}
+
+func (f *Finalizer) WithVoid(fi func()) *Finalizer {
+	return f.With(func() error { fi(); return nil })
+}
+
+func (f *Finalizer) With(fi func() error) *Finalizer {
 	if fi != nil {
 		f.lock.Lock()
 		defer f.lock.Unlock()
 
 		f.pending = append(f.pending, fi)
 	}
+	return f
 }
 
-func (f *Finalizer) Close(c io.Closer) {
+func (f *Finalizer) Close(c io.Closer) *Finalizer {
```

Comment:
In this case Close will finalize with the given closer, therefore it can be chained to add multiple CLosers ar other finalization functions. If does not *cloe* the finalizer, this is done by `Finalize`

Reply:
Thanks, I understand now, that's a bit unfortunate naming, the `Close()` usually closes resources and "finalize" them.